lib/ex_doc/retriever.ex

```diff
@@ -115,7 +117,7 @@ defmodule ExDoc.Retriever do
     modules
     |> Enum.map(&get_module(&1, config))
     |> Enum.filter(&(&1))
-    |> Enum.sort(&(&1.id <= &2.id))
+    |> Enum.sort_by(&({&1.group, &1.id}))
```

Comment:
I think we should order them by the group index in `module_groups`. So the first group will be first, etc. Thoughts?

Reply:
By now order was alphabetically by module name, so I feel like having groups sorted alphabetically isn't to bad. Also if you look at the example above in the PR description I've used a less strict regex as last group to catch all phoenix related modules, which weren't caught by the more precise earlier ones. So the order of groups might be useful for "catch-all" cases more so than for sorting order of the groups.

Reply:
But looking at the way groups worked with pages before, they are ordered by the order of first occurrence of a group, which might be a needed feature. Especially if we make the setup of both the same and deprecate the old way of grouping pages we should look out for that.

Reply:
> But looking at the way groups worked with pages before, they are ordered by the order of first occurrence of a group, which might be a needed feature.

Yup, precisely. 

Plus if we give them control, they can keep it alphabetically. :)

Reply:
I'd really like to retain the user-defined "order of matching" as well, because otherwise the regex patterns might have to be more complex in places. Anything like "group all the remaining modules starting with MyAppWeb" won't be easy if the order in the list does determine the frontend order instead of the matching order. Having a regex exclude the matches of possibly multiple regexes can be quite a pain.

But maybe that's a topic for a later enhancement to the implementation.